fix golint, go vet and go fmt

feliixx opened the pull request:
**WIP - Do not merge**
 
This PR is an attempt to fix `golint`, `go vet` and `gofmt` warning of all files in the project ( see #43 for details) 

A few warnings are still raised by `go vet`: 

```
session.go:661: assignment copies lock value to scopy: mgo.Session contains sync.RWMutex
```
   -> not sure how to handle this one 

```
internal/json/decode_test.go:1350: struct field m has json tag but is not exported
internal/json/decode_test.go:1351: struct field m2 has json tag but is not exported
internal/json/tagkey_test.go:56: struct field tag `:"BadFormat"` not compatible with reflect.StructTag.Get: bad syntax for struct tag key
```
  -> part of the tests and intended to be badly formatted. 

I was able to fix all golint warnings, but bad news is that it currently breaks backward compatibility...
`golint` raise warnings about `Id` not being `ID`, and theres many occurence of this in mgo (`ObjectId`, `FindId()` ...) 

This might be fixed with type alias to preserve backward compatibility ? 

Waiting for any feedback on this, 

Feliixx

domodwyer commented:
Hey @feliixx 

I just had a quick look at the mutex thing - `copySession()` is creating a shallow copy of the struct (so same pointers, etc but fields referenced by value are copied). Essentially it's a lazy way of doing:

```
scopy := Session{
    defaultdb: session.defaultdb,
    sourcedb: session.sourcedb,
    ...
    m: sync.Mutex{},
    ...
}
```

So long as this is the only place a session is copied (and it looks like it is, otherwise you'd have more mutex copy errors) the above change should be fine.

I'm aware of the other two - and I really don't like what I'm about to suggest...

* For the struct tags, I think in this case [disabling](https://golang.org/cmd/vet/#hdr-Struct_tags) vet's struct tag checks might be best - we could do some fancy stuff with build tags but it's a bit hard to reason about. 

If the tags are malformed the code won't work, and this should be caught by any unit tests added alongside any new structs with tags.

* For the `Id => ID` conversion - again, sorry! - I'd rather keep backwards compatibility and not run `golint` as part of the CI - apparently it's not really supposed to be used for this purpose anyway (see [here](https://github.com/golang/lint#purpose)).

I don't think type aliasing will be able to help us with the ID thing but I'm definitely open to ideas - we'd have to do it when 1.9 is a little "older" - a lot of people will currently still be using 1.8 and earlier.

As always, genuinely grateful for the time you've put in contributing - **thanks!**

Dom

feliixx commented:
Hey @domodwyer, 

Totally understand that keeping backward compatibility is more important. 

I removed all Id/ID related changes, and fixed the warning in `copySession()` 

So here is a solution : 

  - run `golint ./... | grep -v 'ID' | cat ` so warnings about Id/ID aren't printed, but all other are. If some warnings are found, the build does not fail

  - run `go vet ` on exported packages only (**mgo**, **mgo/bson** and **mgo/txn**). Content of `internal`  would not be tested . If warnings are found, the build fails 

Feliixx

domodwyer commented:
Hey @feliixx 

Both are good solutions for me :) Any idea why tests on 3.0.X always fail? I've retried them a bunch of times.

Dom

feliixx commented:
Hi @domodwyer, 

tests on 3.0.x were failing because the command `drop` doesn't return the same error as 2.6, 3.2 & 3.4 when trying to delete a collection that doesn't exist. 
It might be a bug, couldn' test this with 3.0.15 (last release for the 3.0.x serie). 
I updated the test `TestRepairCursor`, everything should be fine now!

domodwyer commented:
Thanks @feliixx!

domodwyer commented on bson/bson.go:
```diff
@@ -83,13 +83,13 @@ const (
 	BinaryBinaryOld   byte = 0x02
-	BinaryUUIDOld     byte = 0x03
-	BinaryUUID        byte = 0x04
+	BinaryUUIdOld     byte = 0x03
```
I think these got swept up in a find-and-replace - should still be `ID`

domodwyer commented on bson/bson.go:
```diff
@@ -189,17 +189,17 @@ type Raw struct {
-// ObjectId is a unique ID identifying a BSON value. It must be exactly 12 bytes
+// ObjectId is a unique Id identifying a BSON value. It must be exactly 12 bytes
 // long. MongoDB objects by default have such a property set in their "_id"
 // property.
```
Accidental ID/Id replace (url broken)

domodwyer commented on bson/bson.go:
```diff
@@ -447,7 +447,7 @@ var Undefined undefined
 //   0x01 - Function (!?)
 //   0x02 - Obsolete generic.
-//   0x03 - UUID
+//   0x03 - UUId
```
UUID

domodwyer commented on bson/bson_corpus_spec_test.go:
```diff
@@ -740,15 +740,15 @@ func (s *S) TestDBPointer_type__deprecated__String_not_null_terminated(c *C) {
 }
 
-func (s *S) TestDBPointer_type__deprecated__short_OID__less_than_minimum_length_for_field_(c *C) {
+func (s *S) TestDBPointer_type__deprecated__short_OId__less_than_minimum_length_for_field_(c *C) {
```
OID

domodwyer commented on bson/bson_corpus_spec_test.go:
```diff
@@ -740,15 +740,15 @@ func (s *S) TestDBPointer_type__deprecated__String_not_null_terminated(c *C) {
 }
 
-func (s *S) TestDBPointer_type__deprecated__short_OID__greater_than_minimum__but_truncated_(c *C) {
+func (s *S) TestDBPointer_type__deprecated__short_OId__greater_than_minimum__but_truncated_(c *C) {
```
OID

domodwyer commented on bson/bson_corpus_spec_test.go:
```diff
@@ -1372,7 +1372,7 @@ func (s *S) TestObjectId_Random(c *C) {
 }
 
-func (s *S) TestObjectId_OID_truncated(c *C) {
+func (s *S) TestObjectId_OId_truncated(c *C) {
```
OID

domodwyer commented on bson/bson_test.go:
```diff
@@ -580,25 +580,25 @@ func (s *S) TestUnmarshalNilInStruct(c *C) {
 		"BSON has no uint64 type, and value is too large to fit correctly in an int64"},
 	{bson.M{"": bson.ObjectId("tooshort")},
-		"ObjectIDs must be exactly 12 bytes long \\(got 8\\)"},
+		"ObjectIds must be exactly 12 bytes long \\(got 8\\)"},
```
ID

domodwyer commented on bson/encode.go:
```diff
@@ -275,7 +275,7 @@ func (e *encoder) addElem(name string, v reflect.Value, minSize bool) {
 		case typeObjectId:
 			if len(s) != 12 {
-				panic("ObjectIDs must be exactly 12 bytes long (got " +
+				panic("ObjectIds must be exactly 12 bytes long (got " +
```
IDs

domodwyer commented on bson/encode.go:
```diff
@@ -417,7 +417,7 @@ func (e *encoder) addElem(name string, v reflect.Value, minSize bool) {
 			e.addStr(s.Namespace)
 			if len(s.Id) != 12 {
-				panic("ObjectIDs must be exactly 12 bytes long (got " +
+				panic("ObjectIds must be exactly 12 bytes long (got " +
```
Ids

domodwyer commented on gridfs.go:
```diff
@@ -39,6 +39,26 @@ import (
 
+// GridFS stores files in two collections:
+//
+// - chunks stores the binary chunks. For details, see The chunks Collection.
```
`The` => `the`

domodwyer commented on gridfs.go:
```diff
@@ -376,11 +397,11 @@ func (file *GridFile) Id() interface{} {
 // It is a runtime error to call this function once the file has started
 // being written to, or when the file is not open for writing.
-func (file *GridFile) SetId(id interface{}) {
+func (file *GridFile) setId(id interface{}) {
```
Is not exporting this intentional?

domodwyer commented on internal/json/example_test.go:
```diff
@@ -14,14 +14,15 @@ import (
 func ExampleMarshal() {
 	type ColorGroup struct {
-		ID     int
+		Id     int
```
id

domodwyer commented on internal/json/example_test.go:
```diff
@@ -14,14 +14,15 @@ import (
 	}
 	group := ColorGroup{
-		ID:     1,
+		Id:     1,
```
id

domodwyer commented on internal/json/example_test.go:
```diff
@@ -31,7 +32,7 @@ func ExampleMarshal() {
 	os.Stdout.Write(b)
 	// Output:
-	// {"ID":1,"Name":"Reds","Colors":["Crimson","Red","Ruby","Maroon"]}
+	// {"Id":1,"Name":"Reds","Colors":["Crimson","Red","Ruby","Maroon"]}
```
id

domodwyer commented on gridfs.go:
```diff
@@ -411,7 +432,7 @@ func (file *GridFile) ContentType() string {
 }
 
-// ContentType changes the optional file content type.  An empty string may be
+// SetContentType changes the optional file content type.  An empty string may be
```
This is a breaking change for users of GridFile - I agree this is what the name should be though.

Opinions @globalsign ?

feliixx replied:
The method name was already `SetContentType`, this is just a documentation change

tadukurow replied:
Doesn't look like an issue

domodwyer replied:
My bad, thought it was the method that changed

domodwyer commented on internal/sasl/sasl.go:
```diff
@@ -26,7 +26,8 @@ import (
 
-type saslStepper interface {
+// Stepper interface for saslSession
+type Stepper interface {
```
Intentionally exported?

feliixx replied:
yes, as it is returned by `New(username, password, mechanism, service, host string)` method which is exported ( [sasl.go](https://github.com/globalsign/mgo/blob/master/internal/sasl/sasl.go#L53) ) but maybe it shouldn't be ? I'm not familiar with this part of the driver, so can't tell...

tadukurow replied:
I agree the interface should be exported if returned from New.

domodwyer commented on log.go:
```diff
@@ -34,15 +34,15 @@ import (
 // associated with using an interface rather than the type.  Depending on how
 // often the logger is plugged in, it would be worth using the type instead.
-type log_Logger interface {
+type LogLogger interface {
```
Intentionally exported?

domodwyer commented on session.go:
```diff
@@ -44,23 +44,32 @@ import (
+//
+// Relevant documentation on read preference modes:
+//
+//     http://docs.mongodb.org/manual/reference/read-preference/
```
Thanks for adding all the reference URLs! Great addition 👍

domodwyer commented on session.go:
```diff
@@ -2416,7 +2499,7 @@ func (p *Pipe) Iter() *Iter {
 // server. Ensure the connection has been established (i.e. by calling
 // session.Ping()) before calling NewIter.
-func (c *Collection) NewIter(session *Session, firstBatch []bson.Raw, cursorId int64, err error) *Iter {
+func (c *Collection) NewIter(session *Session, firstBatch []bson.Raw, CursorId int64, err error) *Iter {
```
accidental capitalisation (and in the documentation above)

domodwyer commented on session.go:
```diff
@@ -3828,11 +3917,11 @@ func (iter *Iter) Err() error {
 	iter.m.Lock()
-	cursorId := iter.op.cursorId
-	iter.op.cursorId = 0
+	CursorId := iter.op.CursorId
```
Accidental capitalisation

domodwyer commented on session_test.go:
```diff
@@ -1355,20 +1356,20 @@ func (s *S) TestView(c *C) {
 
 	var result struct {
-		ID int    `bson:"_id"`
+		Id int    `bson:"_id"`
```
id

domodwyer commented on session_test.go:
```diff
@@ -1338,13 +1339,13 @@ func (s *S) TestView(c *C) {
 
 	var viewInfo struct {
-		ID       string   `bson:"_id"`
+		Id       string   `bson:"_id"`
```
id

domodwyer commented on socket.go:
```diff
@@ -137,13 +137,13 @@ func (op *queryOp) finalQuery(socket *mongoSocket) interface{} {
 	collection string
 	limit      int32
-	cursorId   int64
+	CursorId   int64
```
accidental capitalisation

domodwyer commented on socket.go:
```diff
@@ -137,13 +137,13 @@ func (op *queryOp) finalQuery(socket *mongoSocket) interface{} {
 type replyOp struct {
 	flags     uint32
-	cursorId  int64
+	CursorId  int64
```
accidental capitalisation

domodwyer commented on socket.go:
```diff
@@ -171,7 +171,7 @@ type deleteOp struct {
 
 type killCursorsOp struct {
-	cursorIds []int64
+	CursorIds []int64
```
accidental capitalisation

domodwyer commented on txn/txn.go:
```diff
@@ -137,7 +137,7 @@ func newNonce() string {
 type token string
 
-func (tt token) id() bson.ObjectId { return bson.ObjectIdHex(string(tt[:24])) }
+func (tt token) Id() bson.ObjectId { return bson.ObjectIdHex(string(tt[:24])) }
```
accidental capitalisation

domodwyer commented on txn/txn.go:
```diff
@@ -207,10 +207,13 @@ func (op *Op) name() string {
 	// Assert value to assert that the document with the given
-	// Id exists or does not exist, respectively.
-	DocExists  = "d+"
+	// Id exists.
```
ID

domodwyer commented on txn/txn.go:
```diff
@@ -207,10 +207,13 @@ func (op *Op) name() string {
+	DocExists = "d+"
+	// DocMissing may be used on an operation's
+	// Assert value to assert that the document with the given
+	// Id does not exist.
```
ID

domodwyer commented on txn/txn.go:
```diff
@@ -291,12 +296,12 @@ var ErrAborted = fmt.Errorf("transaction aborted")
 // Any number of transactions may be run concurrently, with one
 // runner or many.
-func (r *Runner) Run(ops []Op, id bson.ObjectId, info interface{}) (err error) {
+func (r *Runner) Run(ops []Op, Id bson.ObjectId, info interface{}) (err error) {
```
accidental capitalisation

domodwyer commented on txn/txn.go:
```diff
@@ -366,11 +371,11 @@ func (r *Runner) ResumeAll() (err error) {
 // of the Run method after the transaction is inserted.
-func (r *Runner) Resume(id bson.ObjectId) (err error) {
-	t, err := r.load(id)
+func (r *Runner) Resume(Id bson.ObjectId) (err error) {
```
accidental capitalisation

domodwyer commented on txn/txn.go:
```diff
@@ -394,7 +399,7 @@ func (r *Runner) Resume(id bson.ObjectId) (err error) {
 // Saved documents are in the format:
 //
-//     {"_id": <txn id>, <collection>: {"d": [<doc id>, ...], "r": [<doc revno>, ...]}}
+//     {"_id": <txn Id>, <collection>: {"d": [<doc Id>, ...], "r": [<doc revno>, ...]}}
```
accidental capitalisation

domodwyer commented on txn/txn.go:
```diff
@@ -481,11 +486,11 @@ func (r *Runner) PurgeMissing(collections ...string) error {
 }
 
-func (r *Runner) load(id bson.ObjectId) (*transaction, error) {
+func (r *Runner) load(Id bson.ObjectId) (*transaction, error) {
```
accidental capitalisation

domodwyer commented on txn/txn.go:
```diff
@@ -438,7 +443,7 @@ func (r *Runner) PurgeMissing(collections ...string) error {
 					continue
 				}
-				logf("WARNING: purging from document %s/%v the missing transaction id %s", collection, tdoc.Id, txnId)
+				logf("WARNING: purging from document %s/%v the missing transaction Id %s", collection, tdoc.Id, txnId)
```
accidental capitalisation

domodwyer commented on txn/txn.go:
```diff
@@ -467,7 +472,7 @@ func (r *Runner) PurgeMissing(collections ...string) error {
 				continue
 			}
-			logf("WARNING: purging from stash document %s/%v the missing transaction id %s", stdoc.Id.C, stdoc.Id.Id, txnId)
+			logf("WARNING: purging from stash document %s/%v the missing transaction Id %s", stdoc.Id.C, stdoc.Id.Id, txnId)
```
accidental capitalisation

domodwyer commented on txn/txn.go:
```diff
@@ -542,7 +546,7 @@ func valueNature(v interface{}) (value interface{}, nature typeNature) {
 		return v, natureStruct
 	}
-	panic("document id type unsupported by txn: " + rv.Kind().String())
+	panic("document Id type unsupported by txn: " + rv.Kind().String())
```
accidental capitalisation

tadukurow commented on bson/bson.go:
```diff
@@ -99,12 +99,12 @@ type Getter interface {
 }
 
-// A value implementing the bson.Setter interface will receive the BSON
+// Setter interface : a value implementing the bson.Setter interface will receive the BSON
```
The space between "interface :" should be removed for consistency with the Getter comment. Very nitpicky this.

tadukurow commented on bson/bson.go:
```diff
@@ -130,10 +130,10 @@ type Setter interface {
 // its respective zero value. When used in pointer values, this will set the
 // field to nil rather than to the pre-allocated value.
-var SetZero = errors.New("set to zero")
+var ErrSetZero = errors.New("set to zero")
```
Potentially breaks backwards compat for little gain I'd say. @domodwyer, not sure how likely is this to be used?

domodwyer replied:
I don't think this is much of an issue, it's unlikely to be used by 99% of people.

medicalwei replied:
FYI, I am one of the 1% that's affected by this change. Add reactions here for "me too".

tadukurow commented on gridfs_test.go:
```diff
@@ -141,12 +141,12 @@ func (s *S) TestGridFSFileDetails(c *C) {
 	c.Assert(info, IsNil)
 
-	file.SetId("myid")
+	// file.setId("myid")
```
Any reason to keep this around?
